yetanotherco/aligned_layer
# feat(batcher client): add verify function in service manager to check proof merkle path

Closes #320 

A new function has been added to the aligned service manager contract, `verifyBatchInclusion`. This function receives what is responded to the Aligned users when they send their proofs to the batcher and verifies that the proof sent to Aligned is included in a verified batch.

To this end, the batcher client has been refactored to admit two subcommands: `submit`, to send the proofs just as we have been doing up to now, and `verify-proof-onchain`, to use the new contract feature.

To test it in local devnet:
* Start anvil with block time, operator, aggregator and batcher.
* To test it just as a user will use it, run `make install_aligned_compiling`.
* submit proofs with
```
aligned submit --proving_system GnarkPlonkBls12_381 \
--proof batcher/aligned/test_files/plonk_bls12_381/plonk.proof \
--public_input batcher/aligned/test_files/plonk_bls12_381/plonk_pub_input.pub \
--vk batcher/aligned/test_files/plonk_bls12_381/plonk.vk \
--repetitions 4
```
when responses are received from the batcher, each one will be written into a json file inside the `aligned_verification_data` folder with the format `{batch_merkle_root}_{proof_index_in_batch}.json`, for example `19f04bbb143af72105e2287935c320cc2aa9eeda0fe1f3ffabbe4e59cdbab691_0.json`.
* To verify that a proof was included in the batch, run
```
aligned verify-proof-onchain \
--aligned-verification-data aligned_verification_data/19f04bbb143af72105e2287935c320cc2aa9eeda0fe1f3ffabbe4e59cdbab691_0.json \
--private-key-store config-files/anvil.ecdsa.key.json
```
Note that your JSON batch inclusion response may or may not have the same name as this example. If everything works correctly, you should get the response 
```
[2024-06-12T18:09:30Z INFO  aligned] Batch inclusion verification response: true
```
You can try the same with the other responses you received too.

**Bonus points**: You can also try changing some data of these JSON batch inclusion data and test the `verify-inclusion` command. In this case, it should output `false`.

### Holesky testing
To test in holesky, be sure to add the `--chain holesky` argument to the `verify-inclusion` subcommand. For example, 
```
aligned verify-proof-onchain \
--aligned-verification-data aligned_verification_data/681fcf9f64e7451a7f250e7945cef644ed52f6efccaec1a3294595405042009c_0.json \
--private-key-store config-files/anvil.ecdsa.key.json \
--chain holesky
```
